### ISSUE TITLE
SSH bastion connection isn't opening target

### ISSUE THREAD

**Issue by Lockszmith-GH** [xpipe-io/xpipe#69]
I have an SSH connection linked to another SSH connection
```mermaid
flowchart LR
   local["My local machine<br/>Running xpipe"]--o bast["Bastion host"] --> target["Target host"]
```

When attempting to log into Target host, I got the bastion host instead.

When looking at the insights (which, BTW, is an awesome feature) for the target host, I see the ssh command as 'local' connection to the target machine.

Using System Informer, I see the following (sanitized) command line:
```
ssh  -t bast-user@bastion.host -p <custom-bastion-port> -oStrictHostKeyChecking=accept-new -oNumberOfPasswordPrompts=1 -oConnectTimeout=5 ZDOTDIR="/home/username/.xpipe/temp/571941581/zsh" zsh
```
I'm assuming here that you're calling `zsh` and running a zsh script xpipe sent over scp first that chains the conneciton.

Practically following the [accepted answer in this SE Unix Question](https://superuser.com/a/97007/100730).

If so, this would explain it, as the shells on my bastion machine here are 'different'.

May I suggest you use [ProxyJump](https://man.openbsd.org/ssh.1#J) or [`-o "ProxyCommand..."`](https://man.openbsd.org/ssh.1#o) instead - They have served me well, and seem to be a simplified version of these multi-hop connections.

**Comment by Lockszmith-GH** [comment]
Note that for ProxyJump, if you don't want to compose a complex command line, you can compose a temporary ssh-config file and [pass the config file via command line](https://man.openbsd.org/ssh.1#F) to ssh, this allows you to do complex configuraiton on each host, and just chain using the [`ProxyJump`](https://man.openbsd.org/ssh_config.5#ProxyCommand) config entry.

**Comment by crschnick** [comment]
The script generated in `/home/username/.xpipe/temp/571941581/zsh` should connect you to your target host from your bastion, so maybe you can take a look at the contents of it. Or what exactly do you mean by your shells are different?

The reason I don't use any ProxyCommand features or similar is that the current implementation allows for more flexibility as we are not bound to also connect via SSH, you can use any shell connection as proxy.

If you're interested in seeing what is being executed, you don't have to use an external tool. Under settings -> troubleshoot, you can launch it in debug mode. That will show everything, including sensitive information like passwords, so don't share these logs.

**Comment by crschnick** [comment]
Are there any updates on this problem?

**Comment by Lockszmith-GH** [comment]
Not really, (thanks for the excellent troubleshooting feature of debug), and I tried reproducing the issue on my own.

Right now I'll brush it off as a 'me' issue, until I figure out the reason, and whether thers is anything xpipe can do differently here.

**Comment by Lockszmith-GH** [comment]
So, on this particular system, the solution of setting up `ZDOTDIR` var and calling `zsh` doesn't work.
If you would have called `zsh <script file>` it should work.

Is there a particular reason why you rely on ZDOTDIR and not just call the script?

**Comment by crschnick** [comment]
The reason is we try to keep the shell open. If we call `zsh <file>`, the shell will close after it is finished executing the script. Any other shells have some kind of functionality to execute an init script and stay open, but zsh does not, so this is a workaround (an ugly one).

**Comment by crschnick** [comment]
Is there any particular reason why setting ZDOTDIR does not work in this instance though? It worked in all my testing

**Comment by Lockszmith-GH** [comment]
I don't really know yet, I'll continue exploring - it's currently running `zsh 5.8.1 (x86_64-ubuntu-linux-gnu)`
however this machine is an old Ubuntu Jammy install that was highly customized back in the day.

Why rely on `zsh` then? Why not prefere sh, a POSIX shell that will ALWAYS be around?
Is it becuase it's the default for the system?

**Comment by Lockszmith-GH** [comment]
Yes, I see - changing the default shell to `sh` resolved the issue for me.

**Comment by crschnick** [comment]
I just try to honour these system default shell, mainly because that guarantees that the correct profiles are loaded for people who customize their dotfiles. This could be relaxed in theory for intermediate systems to just use sh instead.

**Comment by crschnick** [comment]
One possibility I can see is that the environment variable is not properly passed by ssh, i.e. the command `ssh ... ZDOTDIR="..." zsh` might not get interpreted correctly. We could try what happens if you modify the original command from the the debug log to instead set the variable beforehand manually in a terminal, i.e. executing:
```
export ZDOTDIR="..."
ssh ... zsh
```

**Comment by TableCoffee** [comment]
I'm running into a similar problem but I'm not sure if it's for the same reason. Trying to go from local host - > Jump server - > Target host I only land in my jump server and not the target one.  I see the pushed files in the .xpipe/temp folder in the jump server but unsure why they arent getting executed. My local host default shell is zsh but jump server is bash. I tried turning on debug mode and maybe I just don't understand it but the app closes but doesnt relaunch. Running xpipe on arch via the AUR. What can I do to troubleshoot further?

**Comment by crschnick** [comment]
Ok, but for example the file browser displays the correct data? It is just the terminal launch that is wrong?

About the debug mode, it should reopen a new terminal tab with debug output. If that doesn't work, you can also manually execute `/opt/xpipe/app/scripts/xpiped_debug.sh` .

**Comment by TableCoffee** [comment]
Yep file browser works no problem, displaying the correct files from the target host. 

And thanks, I was looking in the wrong places to figure out the debug mode myself, that works. The debug output is showing normal finish for running the script on the jump server.

**Comment by crschnick** [comment]
Since your jump server is using bash, I will have to rethink possible causes and experiment further here.

**Comment by crschnick** [comment]
Can you provide as much details as possible (without leaking any details of course) of your connections? I guess you can just screenshot the window

**Comment by TableCoffee** [comment]
![Screenshot_20230819_163934](https://github.com/xpipe-io/xpipe/assets/110047644/c065e21d-27ff-482f-adf8-f407543cbc6c)

Does that work?

**Comment by crschnick** [comment]
Yes. And that happens for all connections where the jump server is used?

**Comment by TableCoffee** [comment]
Yes, for the 2 I've tested so far anyway. I haven't connected any others yet.

**Comment by crschnick** [comment]
Ok so I managed to reproduce this, it only happens for me for connections that require a password. As most of mine are key-based, I somehow didn't catch that.

**Comment by TableCoffee** [comment]
Gotcha! Yes all of mine are keys + passphrase. Glad you found it!

**Comment by crschnick** [comment]
Can you check whether this works for you: https://github.com/xpipe-io/xpipe_staging/releases/tag/1.5.3

**Comment by TableCoffee** [comment]
Will do, as fate would have it my motherboard kicked the bucket last night so going to pick one up and rebuild today. Once I'm back in action I'll test it out. I appreciate the help!

**Comment by TableCoffee** [comment]
Ok, back after a rebuild and after configuring the 1.5.3 version identical to 1.5.2 it's properly getting me to the target servers!

**Comment by crschnick** [comment]
So I am closing this as it should be fixed in 1.5.3.
